# 使用pyvogen
首先导入`vogen`库

In [3]:
import vogen

## vogen数据结构
vogen工程的数据结构分为三个层次：`VogNote`、`VogUtt`、`VogFile`。

### VogFile 文件
用vogen.openvog打开一个文件，即返回VogFile对象。

VogFile拥有以下属性：
|属性名|数据类型|说明|
|-|-|-|
|timeSig0|str|节拍，例如"4/4"|
|bpm0|float|曲速|
|accomOffset|int|伴奏起点|
|utts|List[VogUtt]|乐句列表|

In [4]:
#打开文件
vf=vogen.openvog("少女神明.vog")
type(vf)

vogen.VogFile

In [5]:
#查看文件内容
print(vf)

VogFile 4/4 150.0
 VogUtt Gloria man 
  VogNote 57 透[tou] 960 240
  VogNote 60 明[ming] 1200 240
  VogNote 62 的[de] 1440 480
  VogNote 69 空[kong] 1920 480
  VogNote 67 间[jian] 2400 480
  VogNote 65 将[jiang] 2880 240
  VogNote 64 呼[hu] 3120 240
  VogNote 60 吸[xi] 3360 240
  VogNote 57 心[xin] 3600 240
  VogNote 60 跳[tiao] 3840 240
  VogNote 64 断[duan] 4080 240
  VogNote 62 绝[jue] 4320 240
 VogUtt Gloria man 
  VogNote 62 爱[ai] 4800 240
  VogNote 64 或[huo] 5040 240
  VogNote 65 梦[meng] 5280 720
  VogNote 67 想[xiang] 6000 240
  VogNote 64 都[dou] 6240 240
  VogNote 62 消[xiao] 6480 240
  VogNote 60 逝[shi] 6720 240
  VogNote 62 泯[min] 6960 240
  VogNote 57 灭[mie] 7200 960
 VogUtt Gloria man 
  VogNote 57 轮[lun] 8640 240
  VogNote 60 回[hui] 8880 240
  VogNote 62 百[bai] 9120 480
  VogNote 69 转[zhuan] 9600 480
  VogNote 67 往[wang] 10080 480
  VogNote 65 复[fu] 10560 240
  VogNote 64 只[zhi] 10800 240
  VogNote 60 在[zai] 11040 240
  VogNote 57 一[yi] 11280 240
  VogNote 60 念[nian] 11520 240
  VogNote

In [7]:
vf.timeSig0,vf.bpm0

('4/4', 150.0)

In [8]:
# 合成并播放文件
vf.play()

100%|██████████| 8/8 [00:06<00:00,  1.22it/s]


### VogUtt 乐句
“乐句”是vogen架构中一个特殊的数据结构。其他音乐制作软件的数据结构一般是`乐谱`-`音轨`-`工程`三个层次，而Vogen使用的是低于音轨的“乐句”。

之所以有“乐句”这一设计，是因为vogen作为ai引擎，模型以乐句为单位训练和运行。既不是像拼接引擎一样以音符为单位，也不是将全曲一次性输入模型运行（所需要的内存太大）。

但是乐句与乐句又可以重叠，每个乐句可以独立设置音源和语种，这让vogen一定程度支持多轨合唱。Vogen的[官方演示曲](https://space.bilibili.com/169955)就有一部分是多轨纯人声合唱。

> 我个人认为，这一设计不如传统的`音轨`设计。在多轨合唱时，需要给每一轨独立地设置音量，或者分轨导出做后期，而vogen乐句就没办法很方便地分轨导出，要么只能全部合在一起导出，要么就每个乐句单独导出，变成几十个小文件。而Aisingers、X Studio是按休止符自动分句。

> 尽管这样，PyVogen仍然会尽量兼容vogen的各种设计。

VogUtt拥有以下属性：
|属性名|数据类型|说明|
|-|-|-|
|singerId|str|歌手
|romScheme|str|语种（man：中文普通话，yue：粤语，yue-wz：粤语梧州话）
|notes|List[VogNote]|音符列表

### VogNote 音符
VogNote拥有以下属性：
|属性名|数据类型|说明|
|-|-|-|
|pitch|int|音高，C4=60
|lyric|str|歌词汉字
|rom|str|歌词拼音
|on|int|开始时间，1拍=480
|dur|int|时长，1拍=480

## 相关链接
少女神明（节选）
- [本家](https://www.bilibili.com/video/av201393587)
- 作词：芮晴、欢明境
- 作曲：芮晴
- [完整工程下载](https://github.com/oxygen-dioxide/utau-projects/tree/master/%E5%B0%91%E5%A5%B3%E7%A5%9E%E6%98%8E)